app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java

```diff
@@ -6,21 +6,58 @@
 import android.os.Bundle;
 
 import androidx.annotation.NonNull;
-import androidx.annotation.Nullable;
 
 import com.huawei.hms.mlsdk.asr.MLAsrConstants;
 import com.huawei.hms.mlsdk.asr.MLAsrListener;
 import com.huawei.hms.mlsdk.asr.MLAsrRecognizer;
-
 import com.igalia.wolvic.speech.SpeechRecognizer;
+import com.igalia.wolvic.utils.LocaleUtils;
+import com.igalia.wolvic.utils.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
 
 public class HVRSpeechRecognizer implements SpeechRecognizer, MLAsrListener {
     private Context mContext;
     private MLAsrRecognizer mRecognizer;
     private Callback mCallback;
 
+    // Language support depends on the region and English is the only one supported everywhere.
+    // https://developer.huawei.com/consumer/en/doc/development/hiai-Guides/ml-asr-0000001050066212
+    private static final List<String> DEFAULT_SUPPORTED_LANGUAGES = Collections.singletonList(MLAsrConstants.LAN_EN_US);
+    // The first element in these two lists corresponds to "Follow device language"
+    private final List<String> mSupportedLanguages = new ArrayList<>();
+    private final List<String> mSupportedLanguageNames = new ArrayList<>();
+
     public HVRSpeechRecognizer(Context context) {
         mContext = context;
+
+        setSupportedLanguages(DEFAULT_SUPPORTED_LANGUAGES);
+        mRecognizer = MLAsrRecognizer.createAsrRecognizer(mContext);
+        mRecognizer.getLanguages(new MLAsrRecognizer.LanguageCallback() {
+            @Override
+            public void onResult(List<String> list) {
+                setSupportedLanguages(list);
+            }
+
+            @Override
+            public void onError(int i, String s) {
+            }
+        });
+    }
+
+    private synchronized void setSupportedLanguages(List<String> supportedLanguages) {
+        mSupportedLanguages.clear();
+        mSupportedLanguages.add(LocaleUtils.DEFAULT_LANGUAGE_ID);
+        mSupportedLanguages.addAll(supportedLanguages);
```

Comment:
Is it guaranteed that DEFAULT_LANGUAGE_ID is not included in supportedLanguages? Otherwise we would end up with a duplicated id.

Reply:
Yes, `DEFAULT_LANGUAGE_ID` is just the string `"default"` which is not a valid language/locale.

Reply:
My question was more focused on whether the code prevents that from happening, not only for default but for any other language. I think what I really meant is that mSupportedLanguages should be a Set instead of a List

Reply:
The problem is that at the moment we need a convenient way to access items at random positions in `mSupportedLanguages`. A class like `LinkedHashSet` provides predictable iteration but not random access.

In any case, I don't think that there is a big risk of getting duplicated values. When this class is created, the content of `mSupportedLanguages` is just `{ "default", "en-US" }` while we ask the voice recognition library for its supported locales.

When the voice recognition library returns its list of supported locales, the content of `mSupportedLanguages` becomes `"default"` plus those locales.